### ISSUE TITLE
Template cloning in R4.1 is significantly slower than in R4.0

### ISSUE THREAD

**Issue by GWeck** [QubesOS/qubes-issues#6329]
**Qubes OS version**

R4.1

**Affected component(s) or functionality**

Qube management (via CLI or Qube manager)

**Brief summary**

Cloning template VMs in R4.0 is rather fast - much faster even than copying the file `root.img` manually. The same operation takes about 4 to 10 times as long in Qubes R4.1, although still faster than copying the file manually.

**How Reproducible**

always

**To Reproduce**
Clone a template VM under both R4.0 and R4.0 and measure the time.

**Expected behavior**

The times needed for cloning should be about the same.

**Actual behavior**

For a debian template of size 9 GB, cloning takes about 10 sec in R4.0, but about 45 sec in R4.1 on identical hardware, for a Windows template of size 37 GB, 12 sec on R4.0 and 126 sec on R4.1.

**Additional context**

Does the fast cloning in R4.0 use copy-on-write techniques?

**Solutions you've tried**

Cloning is still the best and fastest way of duplicating VMs, but keeping the speed provided by R4.0 would be phantastic.

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

Aölready mentioned in the community forum in [Fedora Template cannot be cloned](https://qubes-os.discourse.group/t/fedora-template-cannot-be-cloned/1946)

**Comment by rustybird** [comment]
> Cloning template VMs in R4.0 is rather fast - much faster even than copying the file root.img manually.

> Does the fast cloning in R4.0 use copy-on-write techniques?

If reflinks are supported by your storage pool's filesystem, i.e. if it's Btrfs (or XFS created with `-m reflink=1`) then yes, you get constant time/space clones thanks to CoW. I'd guess this is the case for your R4.0 installation, but R4.1 was installed on ext4?

**Comment by iamahuman** [comment]
Which pool driver is in use? LVM? reflink (btrfs/xfs)? file?

**Comment by brendanhoar** [comment]
If LVM: are source/target VMs in the same pool? 

If not LVM, did `dd` blocksize get reverted to a smaller value?

**Comment by GWeck** [comment]
Both R4.0 and R4.1 in my installation use the LVM pool driver, but R4.0 has separate logical volumes for each VM (usually `-private` and `-root` , some also `-back`), while R4.1 only has the 3 LVs `pool00`, `root` and `swap`, which are also present in R4.0. Both R4.0 and R4.1 each use only one pool (`qubes_dom0` on `/dev/sda2` and `qubes_dom000` on `/dev/sdb3`, repectively).

These are the relevant parts of my configuration:

[config1.txt](https://github.com/QubesOS/qubes-issues/files/5781162/config1.txt)

**Comment by rustybird** [comment]
Now I'm confused. The ticket description mentioned "copying the file `root.img` manually" which only exists with the 'file' or 'file-reflink' driver, and the error message in your linked forum post also shows that the 'file' driver is used in the R4.1 installation.

What's the output of `qvm-pool` and `qubes-prefs | grep default_pool` in R4.0 and R4.1 dom0?

**Comment by marmarek** [comment]
> These are the relevant parts of my configuration:
> 
> [config1.txt](https://github.com/QubesOS/qubes-issues/files/5781162/config1.txt)

Existence of "snapshot-(numbers)" devices indeed suggests usage of "file" pool driver. But I see there also proper lvm volumes. Make sure your VMs use "lvm" pool, which should be the default. You can list what is where using `qvm-volume` command.

**Comment by GWeck** [comment]
I am somewhat at a loss: I just installed R4.1 from the alpha iso, using mainly the defaults provided by the installer, and - as far as I rember - did not specify any pool structure. The template VMs for fedora and debian were created by the installer. The resulting structure, as far as it is relevant, is shown here:

[qvm-volume.txt](https://github.com/QubesOS/qubes-issues/files/5783084/qvm-volume.txt)

For me, two questions arise:

- Where / how do I influence the pool structure during installation of Qubes or during creation of a VM?

- Is it possible to change the pool structure later on, moving existing VMs to the new structure? How could this be accomplished?
As shown in the earlier post, the structure in my R4.0 installation is quite different.

**Comment by GWeck** [comment]
@rustybird I did not exactly copy a `root.img` file itself, but just a manually created file of the same size in dom0. 

Here is the configuration in R4.0:

[qvm-volume40.txt](https://github.com/QubesOS/qubes-issues/files/5783175/qvm-volume40.txt)

Could it be that the pool `varlibqubes` in R4.1 (see above) is not LVM? If so, what is it, and why?

**Comment by iamahuman** [comment]
`varlibqubes` is the name of the default `file` pool located at `/var/lib/qubes`.

What's the output of `qvm-pool` in your R4.1 installation?

**Comment by iamahuman** [comment]
If `qvm-pool` shows another pool which is indeed LVM and not file, please try cloning the VM to the LVM pool by using:

```sh
qvm-pool -P <LVM pool name> <original vm> testvm-lvm
```

Then, try cloning again but this time on the new VM backed by the LVM pool:

```sh
qvm-pool testvm-lvm testvm-lvm-2
```

If performance regression cannot be observed in the latter, I suppose we can close this issue as not an issue (and perhaps open another ticket as to how the R4.1 installer came to use the file pool driver).

**Comment by GWeck** [comment]
There was no other pool except `linux-kernel`, so I created a new pool `lvm` as LVM pool. Cloning a Windows VM with 27 GB to the LVM pool took several minutes, as I expected, but cloning this VM in the PVM pool was a matter of seconds, like in R4.0.

The fedora-32 template VM cannot be clonded - the `file` driver still complains about a dirty (root?) volume. So I suppose that the VM may be corrupted.

I am now going to reinstall Qubes R4.1 on this test machine; perhaps I'll see what causes the installation to use the `file` driver instead of the `lvm` driver, but it will take some time, because first I have to backup some VMs.

So the original ticket could probably be closed, and we'll see if we need a new, different one.

**Comment by GWeck** [comment]
I just reinstalled R4.1 from the alpha iso of last October, selecting manual partitioning with LVM thin as default and for the `/` and `swap` partitions. Still, the system was created with `varlibqubes` as pool, so it uses the `file` driver.

Now, i am going to repeat this with LVM as default.

Is there any option where I can force using LVM instead of `file`?

**Comment by marmarek** [comment]
At post-installation setup you should have an option to choose specific LVM thin pool: bottom part of this screen: https://openqa.qubes-os.org/tests/12699#step/firstboot/5

I see you do have some existing LVM thin pools (in a config1.txt file you attached earlier) which may interfere with automatic selection.

**Comment by GWeck** [comment]
With LVM, I still get `varlibqubes` and the `file` driver.

In post-installation, I did not select any of the options of `Advanced configuration`.

The LVM thin pools are of a Qubes R4.0 installation on a separate disk `sda2 VG qubes_dom0` , which hopefully is separated from the R4.1 installation on `sdb3 VG qubes_dom000`, but could it be that they collide anyhow?

**Comment by GWeck** [comment]
I I use the advanced options in post-installation, the installer offers just the voulme group `qubes_dom0`, which is from the other system, and not `qubes_dom000`of its own installation. So maybe it gets confused by the two installations existing in parallel?

**Comment by marmarek** [comment]
That is possible... Did you installed (this try) with LVM thin (not just LVM)? Can you confirm you still see some LVM pool created in `qubes_dom000` VG (check on `lvs` output) ?

**Comment by GWeck** [comment]
My current test is with LVM, and I don't see an LVM pool in `qubes_dom000`.

I'll try now with LVM thin.

**Comment by GWeck** [comment]
If I select LVM thin during phase 1 of the installation, I have 3 LVs (pool00, root00 and swap00) in `qubes_dom000`, and post-installation offers me both volume groups in its advanced options, but the wrong one first.

I'll select now `qubes_dom000`, i.e. the newly created one manually and excute post-installation, and see what happens.

**Comment by GWeck** [comment]
Now, the VMs are created in `qubes_dom000`, and `qvm-pool` shows `pool00` as `lvm_thin`.

If I clone a VM, say `fedora-32`, it takes only some seconds, so I suppose that is is now using the LVM driver and not the `file` driver.

The cofiguration file for post-installation, `/root/anaconda-ks.cfg`, shows links to the correct disk `sdb` and to the correct volume group `qubes_dom000`. (There is still the problem from issue #5003 that non-integer values for the sizes of the boot and the root partition were written by phase 1 of the installation. Ironically, the size of the swap partition is correct.)

**Comment by GWeck** [comment]
It boils down to the problem that, if there are two installations of Qubes on the same machine, the post-installation sees both their pools and per default may select the wrong one - maybe in alphabetical sequence.  It should only see the pool on the installation media of the newly created installation or, at least, select that one per default. The information which pool is on the new installation disk is available as shown by the commands `qvm-pool`, `pvscan` and `vgscan`, together with the contents of the parameter file `/root/anaconda-ks.cfg`.

**Comment by marmarek** [comment]
Normally, in addition to having different pool names, different installations are encrypted separately, so one wouldn't see the other. But it seems you hit a corner case with unencrypted setups.
BTW, is non-integer partition size issue (#5003) applicable with (LUKS) encrypted disk too, or only unencrypted?

**Comment by GWeck** [comment]
I did not check this with encryption, which I do not use on this machine containing no sensitive data (mainly a test environment and some presentaions for teaching).

I now started the installation selecting encryption, but not completing it, since at the moment I cannot afford to lose the system on this disk. But I observerd the same behaviour in the installer with and without selecting encryption, so I suppose it is independent from that selection:

- If I specify 1 GiB for the `/boot` partition (standard), this gets changed to 1023,98 MiB, and the root partition (LVM) size gets changed similarly by the installer. I suppose that the installer writes these changed values into `/root/anaconda-ks.cfg`, causing the trouble later on.

- The swap partition retains the specified size of 20 GiB.

**Comment by marmarek** [comment]
Here is the test ISO build with the fix: https://openqa.qubes-os.org/tests/15746/asset/iso/Qubes-4.1-20210113-x86_64.iso
It also includes the workaround for #5003.
Can you give it a try?

**Comment by GWeck** [comment]
Having a new SSD - identical to the old one, I now checked the R4.1 version. The installation created a new pool `pool00` on `qubes_dom000` and selected that for the VM creation. So all VMs got created on a volume with LVM thin provisioning by default, and VM cloning is fast again. This problem is fixed then.

The second phase of the installation still did not start automatically, because `/root/anaconda-ks.cfg` still contains the non-integer values for the partition sizes of the boot and root partitions. **But:** The file `/root/initial-setup-ks.cfg` contains the correct integer values. So, somthing must be happening on the way from there to the other file.

Here are the two files: 
[anaconda-ks.zip](https://github.com/QubesOS/qubes-issues/files/5824823/anaconda-ks.zip)

I'm going to repeat the test for the R4.0 version and report in #5003.

**Comment by marmarek** [comment]
It seems I forgot to actually apply the patch :facepalm:
This one should be better: https://openqa.qubes-os.org/tests/15857/asset/iso/Qubes-4.1-20210117-x86_64.iso

**Comment by GWeck** [comment]
Now it's all o.k.! The file `/root/anaconda-ks.cfg` contains only integer values, and post-installation starts automatically. Also, the newly created VMs are put into the LVM thin pool `qubes_dom000`, such that template cloning is fast again. So finally the cow has left the ice, as we are saying in Germany!

If you wpuld like, I could also test the R4.0 version, now that I have enough disk space.

**Comment by marmarek** [comment]
Thanks for testing! Fresh 4.0 build is here: https://openqa.qubes-os.org/tests/overview?distri=qubesos&version=4.0&build=20210117-4.0&groupid=1 - wait for some of them to turn green and try.

**Comment by GWeck** [comment]
Now that everything is green, is somewhre an iso to test?

**Comment by rustybird** [comment]
If you click on one of the green dots, the .iso file is under the "Logs & Assets" tab.

**Comment by GWeck** [comment]
Thanks a lot - I am downloading it now (will take some time, because my internet connection is lousy).

**Comment by GWeck** [comment]
The test with R4.0 was successful, too. The file `/root/anaconda-ks.cfg` contains only integer values, and post-installation starts automatically. Also, the newly created VMs are put into the LVM thin pool qubes_dom000, such that template cloning is fast as it was before for R4.0. 

So I guess we can close this isse now, as well as #5003.

I appreciate the way we worked together on these isssues very much!

**Comment by andrewdavidwong** [comment]
Closing this as resolved. If you believe the issue is not yet resolved, or if anyone is still affected by this issue, please leave a comment, and we'll be happy to reopen this. Thank you.
